rl_chain/slates_chain.py

```diff
@@ -47,49 +47,107 @@ def __init__(self, actions, label=None):
         self.actions = actions
         self.label = label
 
-    @property
-    def vwtxt(self):
-        context = [f'slates shared {-self.label.r if self.label else ""} |']
+
+class SlatesTextEmbedder(base.Embedder):
+    """
+    Slates Text Embedder class that embeds the context and actions and slates into a format that can be used by VW
+    
+    Attributes:
+        embeddings_model (SentenceTransformer, optional): The type of embeddings to be used for feature representation. Defaults to BERT Sentence Transformer
+    """
+
+    def __init__(self, model: Optional[Any] = None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if model is None:
+            model = SentenceTransformer("bert-base-nli-mean-tokens")
+
+        self.model = model
+
+    def featurize(self, raw_actions: Dict[str, List[str]]):
+        # Build action embeddings
+        actions = []
+        actions_map = []
+        for (k, v) in raw_actions.items():
+            actions.append(v)
+            actions_map.append(k)
+
+        def _str(embedding):
+            return " ".join([f"{i}:{e}" for i, e in enumerate(embedding)])
+
+        action_features = [
+            [
+                _str(self.model.encode(action.impl))
+                if isinstance(action, base._Embed)
+                else action.replace(" ", "_")
+                for action in slot
+            ]
+            for slot in actions
+        ]
+
+        self.action_features = action_features
+        return actions, actions_map, action_features
+
+    def to_vw_format(self, inputs: Dict[str, Any]) -> str:
+        slates_label = inputs.get("slates_label", None)
+
+        context = [f'slates shared {-1.*slates_label.r if slates_label else ""} |']
+
         actions = chain.from_iterable(
             [
                 [f"slates action {i} |Action {action}"]
-                for i, slot in enumerate(self.actions)
+                for i, slot in enumerate(self.action_features)
                 for action in slot
             ]
         )
         ps = (
-            [f"{a}:{p}" for a, p in self.label.ap()]
-            if self.label
-            else [""] * len(self.actions)
+            [f"{a}:{p}" for a, p in slates_label.get_actions_and_probs()]
+            if slates_label
+            else [""] * len(self.action_features)
         )
         slots = [f"slates slot {p} |" for p in ps]
-        return "\n".join(
-            list(chain.from_iterable([context, actions, slots]))
-        )  # TODO: remove
+        return "\n".join(list(chain.from_iterable([context, actions, slots])))
 
 
 class Policy(ABC):
     @abstractmethod
-    def predict(self, decision: Decision) -> Label:
+    def predict(
```

Comment:
something is wrong with this interface:
- decision and inputs are kind of the same (decision is inputs preprocessed by text_embedder)
- having text_embedder (which main goal is actually produce data in vw format) for non-vw policies seems wrong

Looks like VwPolicy should have SlatesTextEmbedder as field but not as input parameter for this method.

Reply:
Looks like predict should have the only argument which can be either decision or inputs.
Original idea was:
inputs - raw input in langchain format
decision - inputs converted to slates optimization format without vw featurization
Given that, it still feels right to keep decision as predict argument (otherwise inputs -> decision conversion will be inevitably copy pasted across implementations). It will also allow not to introduce "slates_input" to inputs.

Reply:
removed decision and any extra state in text embedder

we can think of a different name for text embedder but I think the code is moving towards where we want it